linkerd/duplex/src/lib.rs

```diff
@@ -85,81 +97,170 @@ where
             is_shutdown: false,
             io,
             direction,
+            flushing: false,
         }
     }
 
-    fn copy_into<U>(
+    /// Reads data from `self`, buffering it, and writing it to `dst.
+    ///
+    /// Returns ready when the stream has shutdown such that no more data may be
+    /// proxied.
+    fn copy_into<U: AsyncWrite + Unpin>(
         &mut self,
         dst: &mut HalfDuplex<U>,
         cx: &mut Context<'_>,
-    ) -> Poll<Result<(), io::Error>>
-    where
-        U: AsyncWrite + Unpin,
-    {
+    ) -> io::Poll<()> {
         // Since Duplex::poll() intentionally ignores the Async part of our
         // return value, we may be polled again after returning Ready, if the
         // other half isn't ready. In that case, if the destination has
         // shutdown, we finished in a previous poll, so don't even enter into
         // the copy loop.
         if dst.is_shutdown {
+            // A prior shutdown may not have completed, so continue driving the
+            // shutdown.
             trace!(direction = %self.direction, "already shutdown");
-            return Poll::Ready(Ok(()));
+            return Pin::new(&mut dst.io).poll_shutdown(cx);
+        }
+
+        // If the last invocation returned pending while flushing, resume
+        // flushing and only proceed when the flush is complete.
+        if self.flushing {
+            ready!(self.poll_flush(dst, cx))?;
         }
+
+        let mut needs_flush = false;
         loop {
-            ready!(self.poll_read(cx))?;
-            ready!(self.poll_write_into(dst, cx))?;
-            if self.buf.is_none() {
-                trace!(direction = %self.direction, "shutting down");
-                debug_assert!(!dst.is_shutdown, "attempted to shut down destination twice");
-                ready!(Pin::new(&mut dst.io).poll_shutdown(cx))?;
-                dst.is_shutdown = true;
-
-                return Poll::Ready(Ok(()));
+            // As long as the underlying socket is alive, ensure we've read data
+            // from it into the local buffer.
+            match self.poll_buffer(cx)? {
+                Poll::Pending => {
+                    // If there's no data to be read and we've written data, try
+                    // flushing before returning pending.
+                    if needs_flush {
+                        // The poll status of the flush isn't relevant, as we
+                        // have registered interest in the read (and maybe the
+                        // write as well).
```

Comment:
might be worth also mentioning that `self.poll_flush` sets the flushing flag so we'll continue the flush if necessary